### ISSUE TITLE
How to detect Transport timeouts in (RPC) client?

### ISSUE THREAD

**Issue by veger** [micro/go-micro#878]
I am using the (default) HTTP transport on an RPC client (so rpc_codec.go is used).

I want to timeout receiving messages as it locks the client mutex, so I cannot send (I need/use bidirectional stream), which works. After a while an error is returned (instead of indefinitely blocking the `Recv()` call like before).

Unfortunately it is impossible to properly detect there was a timeout: [rpc_coded.go#227](https://github.com/micro/go-micro/blob/master/client/rpc_codec.go#L227) converts the transport (timeout) error into a string and puts it in `Error.Details`!
So I would need to parse the string...?!

I don't think this is very nice, so I wonder if I am using the timeout (detection) wrongly?

**Comment by veger** [comment]
I guess I am circumventing #532, a proper fix would not require to detect the Timeouts (as I do not want/need the timeouts on `Recv()` at all)

**Comment by asim** [comment]
If there's a timeout the context is cancelled. You can check context.Done()

**Comment by veger** [comment]
There is no context in [`Recv()`](https://github.com/micro/go-micro/blob/master/client/rpc_stream.go#L75)
And [HTTPTransport](https://github.com/micro/go-micro/blob/master/transport/http_transport.go#L141) also does not seem to have/use contexts. It just sets deadline and read response

```
	if h.ht.opts.Timeout > time.Duration(0) {
		h.conn.SetDeadline(time.Now().Add(h.ht.opts.Timeout))
	}

	rsp, err := http.ReadResponse(h.buff, r)
	if err != nil {
		return err
	}
```

But the whole point of me trying to set timeout is to be able to send/receive on the same RPC stream, which is not possible due to the Mutex in `Send()` and `Recv()`...

**Comment by veger** [comment]
For us this issue is obsolete, we needed to detect the Timeouts to fake 'true bidirectional streaming', implemented by #884.
So feel free to close (unless you are interested to providing ways to detect these Timeouts as well)

**Comment by vtolstov** [comment]
@asim so now close the issue as we now have grpc by default?
